### PR TITLE
feat(lib): add missing parts of lang.xml

### LINKED ISSUE #579: Migrate `lang.xml`

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose

Complete the lang.xml functionality needed by XML navigation expressions.

## Approach

- Implement the missing lang.xml operations and runtime behavior.
- Cover successful operations, callback validation, and panic conditions with corpus tests.

## Checklist

- [ ] Read the [Contributing Guide](CONTRIBUTING.md)
- [x] Added or updated [corpus tests](AGENTS.md#corpus-tests)

## Remarks

Closes #579.
Part of #290.
Depends on #872.

## Summary by CodeRabbit

* **New Features**
  * Expanded XML APIs for construction, parsing, traversal, filtering, mapping, iteration, slicing, and transformation.
  * Added namespace-aware names, attributes, serialization, expanded-name access, and typed iterators.
  * Added validation for XML elements, text, comments, processing instructions, namespaces, and characters.
  * Added support for named arguments when retrieving XML items.
  * Improved XML sequence normalization, concatenation, child replacement, and cycle detection.

* **Bug Fixes**
  * Improved namespace handling, equality comparisons, invalid-content reporting, and navigation-bound validation.

* **Tests**
  * Added extensive coverage for XML APIs, callbacks, namespaces, errors, iterators, named arguments, and edge cases.

### CODE REVIEW

File: test_util/skip.go [hunk 1]
```diff
@@ -52,6 +52,11 @@ var WASMUnsupportedTests = []string{
 // "subset8/08-foo/bar-v.bal"). They are matched as suffixes against either a
 // relative test name or an absolute path.
 var UnsupportedTests = []string{
+	// Empty XML is represented as XMLText at runtime, so its iterator loses the
+	// statically inferred xml<never> constituent type at the extern boundary.
+	// https://github.com/ballerina-nutcracker/ballerina/issues/853
+	"subset10/10-xml-navigation/empty-iterator-static-type-v.bal",
+
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply test_util/skip.go:59]: [AI RESPONSE] ignoring due to this being a deliberate, tracked deferral rather than an oversight. The diagnosis is accurate — `execNewXMLText` builds the value unconditionally and `xmlIterator` derives the iterator type from `xmlValue.Type()`, so an empty `xml<never>` yields an `xml:Text` iterator — but it is already recorded in issue #853 ("Preserve monomorphized XML iterator type at native boundary"), which the skip entry links. The real fix is a BIR/native-call contract change so monomorphized result types reach native implementations; that is architectural and out of scope for a lang.xml completion PR. jBallerina 2201.13.4 also prints `false` here, so this is not a regression introduced by this PR. The skip follows the established `UnsupportedTests` convention: explanatory comment plus tracking issue.

File: semantics/internal/types/type_resolver.go [hunk 2]
```diff
@@ -8741,6 +8755,112 @@ func createXMLIteratorType(t typeResolver, itemTy semtypes.SemType) semtypes.Sem
...
+	containerTy, itemTy, chain, ok := resolveXMLOpaqueContainer(t, chain, args, pos)
+	if !ok {
+		return model.SymbolRef{}, chain, false
+	}
+	if sym.Lookup != nil {
+		if ref, found := sym.Lookup(containerTy); found {
+			return ref, chain, true
+		}
+	}
+	sig := model.TypedFunctionSignature{ParamTypes: []semtypes.SemType{containerTy, semtypes.Int, semtypes.Int}, RestParamType: semtypes.Never, ReturnType: semtypes.XMLSequence(itemTy), Flags: model.FuncSymbolFlagIsolated}
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semantics/internal/types/type_resolver.go:9032]: [AI RESPONSE] ignoring due to a deliberate scope decision: for opaque functions we support named arguments but not defaultable parameters. The analysis is correct — `x.slice(1)` is valid per the spec and currently fails with `missing required parameter 'endIndex'` — but defaultable-parameter support for opaque symbols is a separate concern from parameter naming, and we are not extending it here. Note that named arguments on `slice` already work, since `opaqueFunctionParams` carries its `x`/`startIndex`/`endIndex` names; only the short arity is unsupported. Tracking the short-arity form separately rather than bundling it into this PR.

File: semantics/internal/types/type_resolver.go [hunk 3]
```diff
@@ -8441,6 +8446,15 @@ func opaqueFunctionParams(pkg, name string, sig model.TypedFunctionSignature) []
...
 			return []model.Param{{Name: "m"}, {Name: "k"}}
 		}
+	case "lang.xml":
+		switch name {
+		case "iterator":
+			return []model.Param{{Name: "x"}}
+		case "slice":
+			return []model.Param{{Name: "x"}, {Name: "startIndex"}, {Name: "endIndex"}}
+		case "map", "forEach", "filter":
+			return []model.Param{{Name: "x"}, {Name: "func"}}
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semantics/internal/types/type_resolver.go:8652]: [AI RESPONSE] Fixed. Added the `get` case to `opaqueFunctionParams` returning `{{Name: "x"}, {Name: "i"}}`, as suggested. Worth noting the impact was slightly broader than described: the fallthrough returned all-unnamed params, so *no* named argument resolved on `xml:get` — `xml:get(x = x, i = 1)` failed too, not just the `i` form. Covered by a new corpus test `corpus/bal/subset10/10-xml-navigation/xml-get-named-arg-v.bal` exercising `x.get(i = 1)`, `xml:get(x, i = 0)` and `xml:get(x = x, i = 1)`; output verified against jBallerina.

File: semantics/internal/types/type_resolver.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semantics/internal/types/type_resolver.go:9032]: [AI RESPONSE] ignoring due to this being a known limitation that is tracked separately, not a defect in this PR.

The observation is accurate — per spec `slice(xml<ItemType> x, int startIndex, int endIndex = x.length())`, so `x.slice(1)`, `xml:slice(x, 1)` and `x.slice(startIndex = 2)` are all valid and are currently rejected with `missing required parameter 'endIndex'`.

But this is not specific to `xml:slice`. `slice` is an opaque (typeParam) function, and opaque symbols carry no signature of their own, so defaultable parameters are unsupported for the whole class of them. `padArgTypesForDefaults` (`semantics/internal/types/defaultable_params_handler.go:34`) bails out explicitly:

```go
if _, ok := sym.(*model.OpaqueFunctionSymbol); ok {
    // TODO: there are type param functions with default params need to think how to implement this
    lookup.unimplemented("generic functions with default params not implemented", loc)
    return argTys
}
```

Adding defaultable-parameter support for opaque functions is out of scope for this PR. It is tracked by #910 ("[New Feature]: Support defaultable parameters in langlib functions backed by opaque symbols"), and the proper fix is attaching real untyped signatures to opaque symbols rather than special-casing `slice`.

Working around it per-function would mean a bespoke arity-variant path for each affected langlib function, which is the thing #910 exists to avoid.

File: lib/langlibs/go/lang.xml/xml.go [hunk 4]
```diff
@@ -29,45 +32,367 @@ const (
...
+		}
+		if aURI == "" && aLocal == "xmlns" {
+			panic(values.NewErrorWithMessage("xmlns must be a namespace attribute"))
+		}
+		attrs = append(attrs, values.MapEntry{Key: key, Value: value})
+	}
+	return values.NewXMLElement(ctx.TypeCtx(), "", local, uri, values.NewXMLStringMap(ctx.TypeCtx(), false, attrs), values.NewXMLStringMap(ctx.TypeCtx(), false, namespaces), values.NewNormalizedXMLSequence([]values.XMLValue{child}), false), nil
+}
+
+func validateNamespace(prefix, uri string) error {
```
Comment (warunalakshitha) [inline lib/langlibs/go/lang.xml/xml.go:255]: xmlnames have validateNamespaceBinding. can't we use that?

**Comment by warunalakshitha** [comment]
**PR Review Result**

Code review report below, generated per the process described in [discussion #988](https://github.com/ballerina-nutcracker/ballerina/discussions/988). This process is still being evaluated, so the report may contain false positives — please address each finding, or reply with your findings/defense. If you spot a problem with the report itself, raise it in the discussion thread.

**Subject:** PR #873 review — 3 Important findings, fix first (89% patch coverage)

3 Important findings — fix first. Both are fixable without touching test goldens: `getAttributes()` needs to return a live view (confirmed against jBallerina), a namespace-validation helper was reimplemented instead of reused, and `INTENT.md` needs to exist and be developer-authored (it was missing on this branch, so this run drafted one — it must be rewritten in the author's own words before the PR opens).

---

---
issue: 873
reviewed_diff: sha256:3f8bd91d738d41447efedd584d5f20191210e016c4a71f5db744473d269cb55d
skill_version: sha256:dee208f9f7a47a193b4d0bd8c00d0bcfffad31b0de26a78f39e8b33e01464f57
agent: claude-code / claude-sonnet-5
date: 2026-09-24T03:40:18Z
patch_coverage: 89
important: 3
verify: 0
verdict: "3 Important findings — fix first"
---

# Review report — #873 feat(lib): add missing parts of lang.xml

## Verdict
**3 Important findings — fix first.** All three are fixable without touching test goldens: `getAttributes()` needs to return a live view (confirmed against jBallerina), a namespace-validation helper was reimplemented instead of reused, and `INTENT.md` needs to exist and be developer-authored (it was missing on this branch, so this run drafted one — it must be rewritten in the author's own words before the PR opens).

## Must fix (Important)
| # | Pass | File:line | Issue | Fix |
|---|---|---|---|---|
| 1 | Correctness (jBallerina parity) | `lib/langlibs/go/lang.xml/xml.go:133` `getAttributes()` | Builds a fresh copy of the attribute/namespace entries into a new map. Confirmed against jBallerina (`GetAttributes.java:43`, `return xmlVal.getAttributesMap()`): mutating the returned `map<string>` mutates the live element there, but is a no-op here. Repro: `xml:Element e = xml \`<a x="1"/>\`; map<string> m = e.getAttributes(); m["x"] = "2"; io:println(e);` → jBallerina `<a x="2"/>`, Nutcracker `<a x="1"/>` | Have `getAttributes()` return a map backed by the element's own `Attributes`/`Namespaces` storage (or otherwise alias it) instead of copying entries into a new `NewXMLStringMap`. |
| 2 | Design (reuse before reimplement) | `lib/langlibs/go/lang.xml/xml.go:255` `validateNamespace(prefix, uri)` | Reimplements `values/xml_names.go:113` `validateNamespaceBinding(prefix, uri)` — same four rules (reserved `xmlns` prefix/URI, `xml` prefix↔URI pairing, prefixed-with-empty-URI), but drops two of its checks (`isNCName(prefix)`, `ValidateXMLCharacters(uri)`). Not currently exploitable: `createElement`'s call site already guarantees both preconditions upstream (`ParseExpandedXMLName` enforces NCName on the prefix; `value` already passed `ValidateXMLCharacters` a few lines earlier). Root cause: `validateNamespaceBinding` is unexported, so this different package couldn't call it. Fragile for any future caller of `validateNamespace` without those same preconditions. | Export `validateNamespaceBinding` (or a wrapper) from `values` and have the langlib call it instead of the local reimplementation; delete `validateNamespace`. |
| 3 | Process | `docs/changes/v0.7.0/873/INTENT.md` | Missing on this branch | This run drafted it from the diff (see file). Per the pre-pr-review process, rewrite Problem/Approach in your own words before opening the PR — an agent-drafted INTENT.md is itself an Important finding. |

## Verify (not counted — run these)
None.

## Flagged for the reviewer (not counted)
| Kind | Detail |
|---|---|
| Dead code | `values/xml.go:117` `writeXMLStringMap` — its only two call sites were removed when `XMLString()` was rewritten to use the new `values/xml_serialize.go` renderer (`staticcheck` U1000). Delete it. |
| Coverage below target (per-file, not overall) | `birgen/bir_gen.go` 73% (1393-1395, 1436 — both internal-error/impossible-guard paths, see Evidence; no test needed) |
| Coverage below target (per-file, not overall) | `semantics/internal/opaque/registry.go` 76% (403-404, 415-416, 429-430, 441-442: `xml:get`/`xml:slice` "missing argument"/materialize-failure internal errors, likely unreachable from `.bal` — verify before adding tests; 481-483, 487-488, 490-492: `resolveXMLContainer`'s "expect first argument to be a subtype of xml" for `get`/`slice` specifically — the existing `iterator-nonxml-e.bal` covers this message via a different opaque function, not this call site) |
| Coverage gap | `lib/langlibs/go/lang.xml/xml.go` 95% (358-359, 369-370, 381-382): callback-error propagation in `xml:map`/`forEach`/`filter` (`ctx.InvokeFunctionValue` returning an error, e.g. from a `check`-propagating callback) has no corpus test. Also 160: `setChildren` with an invalid-character string panics untested. |
| Coverage gap | `values/xml_names.go` 75%, `values/xml_parse.go` 84%, `values/xml_serialize.go` 93%: remaining uncovered lines are name/character/namespace-binding validation error branches; several already have `-p`/`-e` siblings for the same category (e.g. `lang-xml-name-p.bal`, `lang-xml-namespace-p.bal`) but not every call site is exercised. Worth a pass, not blocking — overall patch coverage is 89%. |
| New dependency | none |
| Corpus fixture changes | 18 `.txtar` files changed, all new goldens for the new functions (no existing behavior changed) |
| Disabled/skipped test | `empty-iterator-static-type-v.bal` added to `test_util/skip.go`, with issue #853 — has a reason and an issue, per team rule |
| Benchmark | not run — this PR doesn't touch `runtime/`'s interpreter loop hot path beyond the new XML functions themselves; `corpus/bench` has no XML-navigation cases to compare against |

## Fixed during review
None — this is a review of an already-open external PR (#873), not a branch this session can push to, so findings are reported rather than fixed. (`patch_coverage.py`'s own module-detection bug, found while running this review, was fixed in the pre-pr-review skill itself — see separate note below.)

## Evidence
| Pass | Result |
|---|---|
| Branch / scope | PR873 head `60c6cfc06` vs merge-base `720728194` (= current `main`); 1 commit, 124 files (+11,961 / −374) |
| INTENT.md | missing — drafted by this run, flagged Important |
| Commits | Conventional ✔ (`feat(lib): add missing parts of lang.xml`) · no Co-authored-by ✔ · no AI author ✔ (single author/committer, human email) |
| Lint, fmt, license headers | `gofmt -l -s` clean; license headers present on all new `.go` files; `golangci-lint`/`staticcheck` not installed locally — ran `go vet` (clean) and `staticcheck@latest` via `go run` instead |
| Dead code | 1 found (`writeXMLStringMap`, orphaned by the `xml_serialize.go` rewrite) — see Flagged |
| `go test` + corpus | `go build ./...`, `go vet ./...` clean; `go test` on all affected packages (values, birgen, runtime, semantics/*, model, ast) pass; `go test ./corpus/...` (full integration + per-stage goldens) pass |
| Impossible-condition scan | Spot-checked `birgen/bir_gen.go`'s new `splitXMLQualifiedName`: the no-colon fallback is unreachable at its one call site (only invoked when a namespace symbol was already resolved, which implies a prefixed name) — harmless, not reported (nit, not user-visible) |
| Sibling consistency | New opaque functions (`get`/`slice`/`map`/`forEach`/`filter`) share `resolveXMLContainer`/`xmlCallbackMonomorphizer` — consistent shape and error messages across all five |
| Reuse | `ConcatXML`/`NewNormalizedXMLSequence` correctly replace the old in-place `XMLText.Body` mutation (a latent aliasing bug in the previous implementation) rather than duplicating it. One miss found: `lib/langlibs/go/lang.xml/xml.go`'s `validateNamespace` reimplements `values.validateNamespaceBinding` instead of reusing it (unexported, so it couldn't) — Important #2 above |
| Mutation check | Not run (external PR, no revert-and-rerun performed) |
| Patch coverage | 89% of 875 changed statement lines (target 80%) — per-file table below. Note: the skill's own `patch_coverage.py` initially misreported 0%/"no statements" for every file in this repo, because `go list -m` (no args) lists every module in this repo's `go.work` workspace instead of just this one — fixed in `.agents/skills/pre-pr-review/scripts/patch_coverage.py` (now uses `GOWORK=off go list -m`) as part of this run. |
| jBallerina parity | 2/3 spot-checked programs match; 1 confirmed divergence (`getAttributes`, Important #1 above). `strip()` (filter comments/PIs → normalize → drop whitespace-only) matches jBallerina exactly, and the prefix-independent `==` fix (`values/equal.go`) is already covered by a corpus test (`lang-xml-errors-v.bal:67-72`, same-URI-different-prefix / different-URI cases) |
| Verification pass | 1 finding checked against jBallerina source and a live repro before being marked Important; the equality-test gap I initially suspected was dropped after finding it's already covered (`lang-xml-errors-v.bal:71-72`) |
| Commands | `go build ./...`, `go vet ./...`, `gofmt -l -s`, `go run honnef.co/go/tools/cmd/staticcheck@latest ./values/... ./lib/langlibs/go/lang.xml/... ./birgen/... ./model/... ./semantics/...`, `go test ./values/... ./lib/langlibs/go/lang.xml/... ./birgen/... ./runtime/... ./semantics/... ./ast/... ./model/... ./desugar/...`, `go test ./corpus/...`, `patch_coverage.py`, `go run ./cli/cmd run` + `bal run` on 3 spot-check programs (`strip`, `getAttributes`, mismatched-prefix equality) |

| File | Changed lines | Covered | Uncovered |
|---|---|---|---|
| ast/expressions.go | 3 | n/a (no statements) | — |
| birgen/bir_gen.go | 15 | 73% | 1393-1395, 1436 |
| lib/langlibs/go/lang.xml/xml.go | 217 | 95% | 160, 213, 227, 231, 358-359, 369-370, 381-382 |
| model/opaque.go | 6 | 100% | — |
| runtime/internal/exec/binary.go | 1 | 100% | — |
| runtime/internal/exec/non_terminators.go | 1 | 100% | — |
| runtime/internal/exec/xml.go | 1 | n/a (no statements) | — |
| semantics/internal/opaque/registry.go | 114 | 76% | 403-404, 415-416, 429-430, 441-442, 481-483, 487-488, 490-492, 507-508, 510-512, 528-529, 545-546, 554-555 |
| semantics/internal/symbols/xmlns_resolver.go | 1 | 100% | — |
| test_util/skip.go | 5 | n/a (no statements) | — |
| values/equal.go | 1 | 100% | — |
| values/xml.go | 123 | 96% | 105-106, 336-337, 373 |
| values/xml_names.go | 71 | 75% | 39-40, 43-44, 56-57, 72-73, 77-78, 92-93, 102-103, 115-116, 118-119 |
| values/xml_parse.go | 129 | 84% | 77, 135-136, 151, 170-171, 173-174, 176-177, 180-181, 201-202, 210-211, 218-219, 227-228 |
| values/xml_serialize.go | 196 | 93% | 75-76, 89, 92, 99, 107, 110, 193-194, 213, 234-235, 238 |

**Comment by warunalakshitha** [comment]
Manually checked the approach and quick scan on code chnages. LGTM. Please address concerns in review report.
